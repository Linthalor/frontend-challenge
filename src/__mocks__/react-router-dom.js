module.exports = {
  ...jest.requireActual('react-router-dom'),
  Navigate: (props) => {
    const MockName = `mocked-react-router-dom-navigate`;
    return <MockName {...props} />;
  }
};