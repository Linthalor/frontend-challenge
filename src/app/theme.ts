import { createTheme } from '@mui/material';

// Themed using Upgrade's palette
export const theme = createTheme({
  palette: {
    primary: {
      light: '#60c7a1',
      main: '#00a669',
      dark: '#008051',
    },
    secondary: {
      main: '#3f3d56',
      light: '#6bc1fe',
    },
    success: {
      main: '#03b57a',
      contrastText: '#ffffff',
    },
    error: {
      main: '#c83c4c',
      contrastText: '#ffffff',
    },
    text: {
      primary: '#3f465e',
      secondary: '#008051'
    }
  },
});