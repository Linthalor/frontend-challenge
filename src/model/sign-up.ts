import { z } from 'zod';
import { t } from 'i18next';

const minNameSize = 1;
const nameMessages = (t('form:name:errors', { returnObjects: true }) as any)(minNameSize);

const minPasswordSize = 8;
const passwordMessages = (t('form:name:errors', { returnObjects: true }) as any)(minPasswordSize);

export const SignUpUserFormValuesSchema = z.object({
  name: z.string({
    required_error: nameMessages.required_error,
  }).min(minNameSize, nameMessages.too_small),
  email: z.string({
    required_error: t('form:email:errors:required_error'),
  }).email(t('form:email:errors:invalid_string')),
  password: z.string({
    required_error: passwordMessages.required_error,
  }).min(minPasswordSize, passwordMessages.too_small),
});
export type SignUpUserFormValues = z.infer<typeof SignUpUserFormValuesSchema>;

// Used to effectively prevent an empty selection.
const minColorSize = 1;

export const SignUpUserPrefsValuesSchema = z.object({
  // TODO: have this fetch the redux's value to compare for validation? might need a component to do that...
  color: z.string({
    required_error: t('form:color:errors:required_error'),
  }).min(minColorSize),
  terms: z.boolean().refine(val => val, {}),
});
export type SignUpUserPrefsValues = z.infer<typeof SignUpUserPrefsValuesSchema>;

export type SignUp = SignUpUserFormValues & SignUpUserPrefsValues;