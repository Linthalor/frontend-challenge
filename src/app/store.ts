import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { signUpReducer } from '../features/sign-up/sign-up.slice';
import { colorReducer } from '../features/color/color.slice';

export const store = configureStore({
  reducer: {
    signUp: signUpReducer,
    color: colorReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
