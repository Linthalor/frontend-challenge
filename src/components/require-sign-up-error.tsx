import { Navigate } from 'react-router-dom';
import { useAppSelector } from '../app/hooks';
import { selectSignUpStatus } from '../features/sign-up/sign-up.slice';
import { isError } from '../model/util/async-prop';

export const RequireSignUpError = ({
  children,
  redirectTo,
}: {
  children: JSX.Element,
  redirectTo: string,
}) => {
  const signUpStatus = useAppSelector(selectSignUpStatus);
  if (isError(signUpStatus)) {
    return children;
  }
  return <Navigate to={redirectTo} replace />;
};