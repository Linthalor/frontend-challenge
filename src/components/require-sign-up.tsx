import { Navigate } from 'react-router-dom';
import { useAppSelector } from '../app/hooks';
import { selectSignUpStatus } from '../features/sign-up/sign-up.slice';
import { isValue } from '../model/util/async-prop';

export const RequireSignUp = ({
  children,
  redirectTo,
}: {
  children: JSX.Element,
  redirectTo: string,
}) => {
  const signUpStatus = useAppSelector(selectSignUpStatus);
  if (isValue(signUpStatus)) {
    return children;
  }
  return <Navigate to={redirectTo} replace />;
};