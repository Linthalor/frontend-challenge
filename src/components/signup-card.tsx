import { Box, Card, CardActions, CardContent, useMediaQuery } from '@mui/material';
import { theme } from '../app/theme';

export const SignUpCard = ({
  content,
  actions
}: {
  content: JSX.Element,
  actions: JSX.Element
}) => {
  const fullscreenSignUpCardMediaQuery = useMediaQuery(theme.breakpoints.down('sm'));
  return (
    <Box
      sx={{
        width: '100%',
        height: '100%',
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
      }}
    >
      <Card
        sx={{
          width: fullscreenSignUpCardMediaQuery ? '100%' : 400,
          height: fullscreenSignUpCardMediaQuery ? '100%' : 'unset',
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'center',
        }}
      >
        <CardContent
          sx={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'stretch',
            justifyContent: 'start',
            gap: 1,
            flex: fullscreenSignUpCardMediaQuery ? 0 : 1,
          }}
        >
          {content}
        </CardContent>
        <CardActions
          sx={{
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'end',
            gap: 1,
          }}
        >
          {actions}
        </CardActions>
      </Card>
    </Box>
  );
};