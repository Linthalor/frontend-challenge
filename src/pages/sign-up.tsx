import { Box, Button, IconButton, InputAdornment, TextField, Typography } from '@mui/material';
import { SignUpCard } from '../components/signup-card';
import { SignUpUserFormValues, SignUpUserFormValuesSchema } from '../model/sign-up';
import { useAppDispatch } from '../app/hooks';
import { selectUserFormValues, setUserFormValues } from '../features/sign-up/sign-up.slice';
import { useNavigate } from 'react-router-dom';
import {zodResolver} from '@hookform/resolvers/zod';
import { useSelector } from 'react-redux';
import { EmailRounded, Visibility, VisibilityOff } from '@mui/icons-material';
import { useForm } from 'react-hook-form';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';

export const SignUp = () => {
  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const { t } = useTranslation();
  const signUpUserData = useSelector(selectUserFormValues);
  const [seePassword, setSeePassword] = useState(false);

  const { handleSubmit, register, formState } = useForm({
    resolver: zodResolver(SignUpUserFormValuesSchema),
    mode: 'onBlur',
    defaultValues: {
      name: signUpUserData?.name || '',
      email: signUpUserData?.email || '',
      password: signUpUserData?.password || '',
    },
  });
  const { errors } = formState;

  const submitForm = (values: SignUpUserFormValues) => {
    dispatch(setUserFormValues(values));
    navigate('/more-info');
  }

  return (
    <form
      style={{
        height: '100%',
        width: '100%',
      }}
      onSubmit={handleSubmit(submitForm)}
    >
      <SignUpCard
        content={<>
          <Typography
            sx={{
              textAlign: 'center',
              fontWeight: 'bold'
            }}
            variant="h5"
            color="text.primary"
            gutterBottom
          >
            {t('page:signUp:label')}
          </Typography>
          <Box sx={{ gap: 2, display: 'flex', flexDirection: 'column' }}>
            <TextField
              id="first-name-field"
              label={t('form:name:label')}
              variant="outlined"
              type="text"
              required
              {...register('name')}
              error={!!errors.name?.message}
              helperText={errors.name?.message}
            />
            <TextField
              id="email-field"
              label={t('form:email:label')}
              variant="outlined"
              type="email"
              required
              {...register('email')}
              error={!!errors.email?.message}
              helperText={errors.email?.message}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <IconButton disabled>
                      <EmailRounded />
                    </IconButton>
                  </InputAdornment>
                ),
              }}
            />
            <TextField
              id="password-field"
              label={t('form:password:label')}
              variant="outlined"
              type={seePassword ? 'text' : 'password'}
              required
              {...register('password')}
              error={!!errors.password?.message}
              helperText={errors.password?.message}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <IconButton onClick={() => setSeePassword(!seePassword)}>
                      {seePassword ? <VisibilityOff /> : <Visibility />}
                    </IconButton>
                  </InputAdornment>
                ),
              }}
            />
          </Box>
        </>}
        actions={<>
          <Button variant="contained" type="submit" disabled={!formState.isValid}>
            {t('common:next')}
          </Button>
        </>}
      />
    </form>
  );
};