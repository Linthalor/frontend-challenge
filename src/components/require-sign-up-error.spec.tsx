import { cleanup, render } from '@testing-library/react';
import { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { createTestStore, store as realStore } from '../app/store';
import { signUpAsync } from '../features/sign-up/sign-up.slice';
import { RequireSignUp } from './require-sign-up';
import { RequireSignUpError } from './require-sign-up-error';
import { SignUpCard } from './signup-card';

jest.mock('@mui/material');
jest.mock('react-router-dom');

describe('RequireSignUpError', () => {
  let store: typeof realStore;
  beforeEach(() => {
    store = createTestStore();
  });
  afterEach(() => {
    cleanup();
  });

  it('should try to redirect without a signup status', () => {
    const route = 'mock-route';
    const MockChildren = 'mock-children';

    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <RequireSignUpError children={<MockChildren />} redirectTo={route} />
        </BrowserRouter>
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display the children with failed signup status', () => {
    const route = 'mock-route';
    const MockChildren = 'mock-children';

    store.dispatch({
      type: signUpAsync.rejected.type,
      error: { message: 'some error' },
    });

    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <RequireSignUpError children={<MockChildren />} redirectTo={route} />
        </BrowserRouter>
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});