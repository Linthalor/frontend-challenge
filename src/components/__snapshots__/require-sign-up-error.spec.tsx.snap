// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequireSignUpError should display the children with failed signup status 1`] = `<mock-children />`;

exports[`RequireSignUpError should try to redirect without a signup status 1`] = `
<mocked-react-router-dom-navigate
  replace="true"
  to="mock-route"
/>
`;
